### PR TITLE
Retrieve minimal set from light client and include supporting headers in client updates

### PR DESCRIPTION
**Pull request by romac**
See: #673

## Description

Retrieve verification trace from the light client and send a `UpdateClient` message for each intermediary header.

## Test

Tested against the Cosmos Hub, by setting the trust threshold to `1/1` in the config, creating a client and updating it shortly afterwards.

<details>
<summary>Click to see example trace</summary>

```Jun 08 14:33:03.346 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264262
Jun 08 14:33:03.348 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264263
Jun 08 14:33:03.349 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264264
Jun 08 14:33:03.351 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264265
Jun 08 14:33:03.352 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264266
Jun 08 14:33:03.354 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264267
Jun 08 14:33:03.355 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264268
Jun 08 14:33:03.356 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264269
Jun 08 14:33:03.358 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264270
Jun 08 14:33:03.359 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264271
Jun 08 14:33:03.360 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for intermediate height 1-264272
Jun 08 14:33:03.362 DEBUG ibc_relayer::foreign_client: [informal-testnet-1 -> cosmoshub-4:07-tendermint-227] MsgUpdateAnyClient for going from trusted height 1-264273 to target height 1-264261
Success: [
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264262,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264263,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264264,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264265,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264266,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264267,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264268,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264269,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264270,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264271,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264272,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
    UpdateClient(
        UpdateClient {
            common: Attributes {
                height: Height {
                    revision: 4,
                    height: 6509134,
                },
                client_id: ClientId(
                    "07-tendermint-227",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 1,
                    height: 264273,
                },
            },
            header: Some(
                Tendermint(
                     Header {...},
                ),
            ),
        },
    ),
]```

</summary>

______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

**Comment by ancazamfir** [comment]
We should also change 
https://github.com/informalsystems/ibc-rs/blob/2b84667d48da53119aa69902cf7e4210e792fc73/relayer/src/foreign_client.rs#L508

and callers to return `Vec<IbcEvents>`

**Comment by romac** [comment]
> We should also change
> 
> https://github.com/informalsystems/ibc-rs/blob/2b84667d48da53119aa69902cf7e4210e792fc73/relayer/src/foreign_client.rs#L508
> 
> and callers to return `Vec<IbcEvents>`

Done in 9f444ea3978f9d65a03a86f2c1a277ccf2aa68eb.

File: relayer/src/light_client/tendermint.rs [hunk 1]
```diff
@@ -110,31 +127,46 @@ impl super::LightClient<CosmosSdkChain> for LightClient {
...
+            let supporting_headers = supporting
+                .into_iter()
+                .map(|h| {
+                    TmHeader {
+                        trusted_height,
+                        signed_header: h.signed_header.clone(),
+                        validator_set: h.validators,
+                        trusted_validator_set: succ_trusted.validators.clone(),
+                    }
+                    .wrap_any()
```
Comment (ancazamfir) [inline relayer/src/light_client/tendermint.rs:159]: shouldn't we call `adjusted_headers()` here also?

**Comment by ancazamfir** [comment]
I did some tests on cosmoshub with updates in the middle with older and newer trusted states. All looks good. 
Not related to this PR but one thing I noticed is that there are no traces coming from the light client and sometimes the `hermes update client` seems to be stuck when it involves hundreds of headers. There is no feedback to the user.
And this also reminds me of the other aspect brought up by DEX about `MsgUpdateClient` not being rewarded at all (even with the fee proposal) and therefore hermes potentially paying a lot for updates on clients, especially on malicious chains.  I think we need to address that in a separate PR.

**Comment by romac** [comment]
I'll add some tracing in the light client on the ibc-rs side, and perhaps we can eventually add some to the tendermint-rs light client as well.

**Review by ancazamfir — APPROVED** [review]
Looks good, great work @romac !!! Thanks!

**Pull request merged**